Add Ruby 3.0 support, and add it to CI

taichi-ishitani opened the pull request:
This PR is to add Ruby 3.0 to CI.

taichi-ishitani committed:
add Ruby 3.0 to CI

taichi-ishitani committed:
add rexml to Gemfile

rexml is a bundled gem from ruby 3.0.
therefore, we need to add rexml to Gemfile to use it via bundler.

saveriomiroddi commented:
Hello!

Thanks for the PR! I have bad news, though - Travis CI are now suspended, as Travis doesn't automatically grant free builds to open source projects.

I've asked information about applying, however, it's not straightforward (in particular, I have several open source projects, and application is per-project).

If you happened to have the interest and the time to convert the build to GitHub actions, that would be super-helpful, and immediately accepted :grimacing: :laughing: :+1:!

taichi-ishitani commented:
Hi @saveriomiroddi ,

> GitHub actions

Sure. I will create the build env on Github actions.

saveriomiroddi commented:
Let me think for a few days, and I'll get back. I've just tried the build on 3.0 locally, and I've seen that it requires `rexml` manual import (and I've noticed your commit).

taichi-ishitani commented:
I invoke `rake spec` on ruby 3.0 and I see no errors.

64kramsystem commented on Gemfile:
```diff
@@ -1,3 +1,7 @@
 
 gemspec
+
+if RUBY_VERSION >= '3.0.0'
```
While this code is working, the rigorous way of comparing versions is by using [`Gem::Version`](https://ruby-doc.org/stdlib-2.7.0/libdoc/rubygems/rdoc/Gem/Version.html).

Check this out:

```ruby
2.7.2 :001 > '3.0' > '10.0'
 => true 
2.7.2 :002 > Gem::Version.new('3.0') > Gem::Version.new('10.0')
 => false
```

:grin: 

When using the Ruby version, there's a ready API:

```ruby
2.7.2 :003 > Gem.ruby_version > Gem::Version.new('10.0')
 => false
```

Although the bad behavior will likely never be triggered, I place much value on exact semantics/logic :smile:.

taichi-ishitani replied:
Thank you for reviewing!
I changed way to check Ruby version according to your proposal.

saveriomiroddi commented:
So! If you change the commit with the more precise version comparison, I will merge this and the other PR!

Thanks for the contributions! :partying_face: :+1:

taichi-ishitani committed:
use Gem's APIs to check Ruby version

(see: https://github.com/saveriomiroddi/spreadbase/pull/18#discussion_r587828375)
